### PR TITLE
Let api compile also detect aliased compilers

### LINKED ISSUE #2477: [REQUEST] Allow to compile with aliased compilers
Currently, aliases need to be resolved to their actual compiler id by the js client.

But we can probably add support for this server-side as well, so raw API calls can be resolved as well.

Maybe there are reasons why we might not want this, but not sure.

### PR DESCRIPTION
**Pull request by partouf**
Fixes #2477

Todo
- [x] Refactor

### CODE REVIEW

File: lib/compilers/fake-for-test.js [hunk 1]
```diff
@@ -30,11 +30,10 @@ export class FakeCompiler {
     }
 
     constructor(info) {
-        this.compiler = {
-            id: info.id || 'fake-for-test',
-            lang: info.lang || 'fake-lang',
-            options: info.options || '',
-        };
+        this.compiler = Object.assign({}, info);
+        this.compiler.id = info.id || 'fake-for-test';
```
Comment (mattgodbolt) [inline lib/compilers/fake-for-test.js:34]: I _think_ you can do this as
`this.compiler = Object.assign({id:'fake-for-test', lang:'fake-lang', options:''}, info);` (if I remember right? the assign from the right param overwrites the left?

File: lib/handlers/compile.js [hunk 2]
```diff
@@ -175,16 +175,28 @@ export class CompileHandler {
...
 
         const langCompilers = this.compilersById[langId];
-        if (langCompilers && langCompilers[compilerId]) {
-            return langCompilers[compilerId];
+        if (langCompilers) {
+            if (langCompilers[compilerId]) {
+                return langCompilers[compilerId];
+            } else {
+                const compiler = _.find(langCompilers, (compiler) => {
+                    return (compiler.compiler.alias) &&
```
Comment (mattgodbolt) [inline lib/handlers/compile.js:183]: super nitpick (and I know you have a TODO to refactor!) but extraneous `()`s I think.

File: test/handlers/compile-tests.js [hunk 3]
```diff
@@ -119,6 +119,36 @@ describe('Compiler tests', () => {
                     });
             });
         });
+
+        it('supports alias compile', () => {
```
Comment (mattgodbolt) [inline test/handlers/compile-tests.js:123]: Thank you _so much_ for adding a test.

File: lib/handlers/compile.js [hunk 4]
```diff
@@ -175,16 +175,28 @@ export class CompileHandler {
...
+
         // If the lang is bad, try to find it in every language
         let response;
         _.each(this.compilersById, compilerInLang => {
             if (response === undefined) {
                 _.each(compilerInLang, compiler => {
                     if (response === undefined &&
-                        (compiler.compiler.id === compilerId || compiler.compiler.alias === compilerId)) {
+                        (compiler.compiler.id === compilerId ||
+                         ((compiler.compiler.alias) &&
```
Comment (mattgodbolt) [inline lib/handlers/compile.js:198]: ditto above :)

**Review by mattgodbolt — APPROVED** [review]
total flyby PR as I was in a boring meeting :grinning:

**Comment by partouf** [comment]
now live

**Pull request merged**